src/storage/src/render/sources.rs

```diff
@@ -575,25 +575,29 @@ fn upsert_commands<G: Scope, FromTime: Timestamp>(
             Some(Ok(ref row)) => match upsert_envelope.style {
                 UpsertStyle::Debezium { after_idx } => match row.iter().nth(after_idx).unwrap() {
                     Datum::List(after) => {
-                        row_buf.packer().extend(after.iter().chain(metadata.iter()));
+                        let mut packer = row_buf.packer();
+                        packer.extend(after.iter());
+                        packer.extend_by_row(&metadata);
                         Some(Ok(row_buf.clone()))
                     }
                     Datum::Null => None,
                     d => panic!("type error: expected record, found {:?}", d),
                 },
                 UpsertStyle::Default(_) => {
                     let mut packer = row_buf.packer();
-                    packer.extend(key_row.iter().chain(row.iter()).chain(metadata.iter()));
+                    packer.extend_by_row(&key_row);
+                    packer.extend_by_row(row);
+                    packer.extend_by_row(&metadata);
                     Some(Ok(row_buf.clone()))
                 }
                 UpsertStyle::ValueErrInline { .. } => {
                     let mut packer = row_buf.packer();
-                    packer.extend(key_row.iter());
+                    packer.extend_by_row(&key_row);
                     // The 'value' record contains all value columns
-                    packer.push_list(row.iter());
+                    packer.extend_by_row(row);
```

Comment:
I think this is a semantics change! `push_list` is not the same as `extend`... it adds an extra level of nesting.

Reply:
Fixed, thanks.

Reply:
You could use `push_list_with` instead, which allows you to access a packer to push the individual list elements. Then, you'd be able to call `extend_by_row` on this packer.